### ISSUE TITLE
IIS -> Lock .NET Core application DLL file

### ISSUE THREAD

**Issue by radenkozec** [dotnet/aspnetcore#3793]
When I try to overwrite .NET Core application DLL file using FTP on production IIS server file is locked and it cannot be overwritten.

To deploy new version I need to stop application in IIS to release lock, then overwrite.
It is not possible to deploy without stopping application.

**Comment by kanadaj** [comment]
@Tratcher @ronnyek  This seems to be related to Application Insights; after adding the App Insights extension, my web app DLL is suddenly very hard to update, no problems before then.

**Comment by dhtek** [comment]
I'm still facing to this problem. I use the flag -enableRule:AppOffline, sometimes it's working well but most of the times deployment is failing saing "ERROR_FILE_IN_USE". 

After starting a release, I monitor the web application folder and see the App_Offline.html dropped by msdeploy. Next the release failed with the above message and note the App_Offline file still in the folder, I start the release again, this times it's working well as the module see the file in the folder.

last asp.net core runtime installed -> aspnetcore.dll 12.1.18263.2

Any ideas ? Still a bug ?

**Comment by poke** [comment]
@dhtek If you try to replace the files immediately after placing the `app_offline.htm`, then the server simply didn't have enough time yet to shut down the application. You should wait a bit before attempting to replace the files.

**Comment by dhtek** [comment]
Ok I understand but I'm using the -enableRule:AppOffline option from msdeploy and the page is automatically dropped it to the folder and the sync is done straight away. I do nothing myself.

**Comment by DM2489** [comment]
I'm stilll seeing this issue too. My project files are referencing `Microsoft.NET.Sdk.Web`, but it is not dropping an `app_offline.htm` file automatically, as stated it should in the docs.

**Comment by markuswolters** [comment]
Hi guys. This annoying issue is still present with Core 2.1 with the lastest VS 2017. I use appoffline, still the main project dll is always is use and requires a complete stop of website during publish. This renders web deploy almost useless.

**Comment by csharmath** [comment]
If you can't take your host offline behind the loadbalancer and would want to do it quicker, here is one way with symlinks and some powershell. 
I was thinking of doing something like this.

It's quicker as you don't have to stop the app pool during the copy + recycle is better than a hard stop as it lets the current requests complete, so less downtime.

```PowerShell
# Setup
# create 2 site root directories
$a = 'C:\inetpub\AspNetCoreSampleA'
$b = 'C:\inetpub\AspNetCoreSampleB'
$siteRoot = 'C:\inetpub\aspnetcoresample'
$siteName = 'AspNetCoreSample'
$poolName = "aspnetcore"
New-Item -Type Directory $a
New-Item -Type Directory $b
# create a symlink to targeting one side
New-Item -Type SymbolicLink -Path $siteRoot -Target $a
# point the site root to the symlink
Set-ItemProperty "IIS:\Sites\$siteName" -name physicalPath -value $siteRoot
# make sure it get's picked up
Restart-WebAppPool -Name $poolName

# this tells you the active side
Get-Item -Path $siteRoot | Select-Object -ExpandProperty target

# Flip the symlink
$current = (Get-Item -Path $siteRoot).Target
$newTarget = if ($current -eq $a) {$b} else {$a}
New-Item -Type SymbolicLink -Path $siteRoot -Target $newTarget -Force
# at this point w3wp.exe still locks the current target folder until it's getting recycled
# Deploy new version to the symlink which is now pointing to the other side which should have no locks
robocopy \\myshare\myapp $siteRoot /mir
# recycle app pool, so it picks up the new files
Restart-WebAppPool -Name $poolName

# bonus point: rollback is easy
$current = (Get-Item -Path $siteRoot).Target
$newTarget = if ($current -eq $a) {$b} else {$a}
New-Item -Type SymbolicLink -Path $siteRoot -Target $newTarget -Force
Restart-WebAppPool -Name $poolName
```

Here is the gist
https://gist.github.com/csharmath/b2af0f50700ce9fbdd8c5c3e582fd41b

**Comment by kanadaj** [comment]
@csharmath or there is also docker I guess. No file locks that way, and it can do rolling updates via swarm/k8s/okd

**Comment by csharmath** [comment]
Sure, that's even better of course. sigh, not every place has that yet.
For the rest of the oldschool setups you need to yakshave it until it works for you with no or minimal downtime.

**Comment by kanadaj** [comment]
@csharmath Agreed, but at this point, given that this has been an unfixed issue for _years_, might as well adapt and go after solutions that work instead of waiting for a miracle...

**Comment by dev8546** [comment]
I have used below power shell script and it is crashed the IIS and even affected the other application pool and I have to restart the machine to get everything back... this process I have got it from the official document and it was disaster , so my advice is DO NOT USE THE app_offline.htm better to use the script which stops the app pool and start it again, one I found in this thread https://github.com/IsraelHikingMap/Site/blob/master/Scripts/Deploy.ps1 

$pathToApp = 'G:\prod_web_core'
$pathToAppOfflineHtml = 'G:\prod_web_core\app_offline.htm'
# Stop the AppPool 
New-Item -Path $pathToApp -Name app_offline.htm
# Provide script commands here to deploy the app
Copy-Item "G:\prod_web_core_temp\*" -Destination $pathToApp -Recurse -Force
# Restart the AppPool 
Remove-Item -Path $pathToAppOfflineHtml
Get-ChildItem -Path "G:\prod_web_core_temp" -Recurse | Remove-Item -Force

**Comment by xsoheilalizadeh** [comment]
I'm force to stop AppPool for success deploy, any progress for fixing this issue?!

**Comment by macfarmw** [comment]
We just encountered this file locking issue for the first time today since dotnet core is new for us. I was really caught off guard by this. It's funny that the number 1 reason I switched from Classic ASP to .Net Framework while it was still in beta (~2001) is because I could do hot deployments with no files locking. It was awesome! Almost 18 years later and now I'm right back where I started. Oh well, you win some you lose some.

**Comment by chrdlx** [comment]
I've just referenced my issue...  in my case deploying Blazor Server Side solutions....
A couple of days ago it was a nightmare... I applied the app_offline.htm, shut down app pools,  and even restarted whole IIS instances to unlock files.  The next step would've been to restart the entire server, luckily after 5 minutes the .dlls unlocked themselves.

**Comment by Portekoi** [comment]
2019, same problem for me... and this issue was created in 2016... still no solution?

**Comment by bladefist** [comment]
2020, same problem for me.  Biggest regression ever.  .Net 4.6 you replace the dll files and the app reloads on the new version.  .Net core locks everything.

**Comment by kanadaj** [comment]
@bladefist Your best bet as of today is to use Docker containers :/

**Comment by bladefist** [comment]
@kanadaj thanks but Dockers is another tier of headaches to deal with.  I eventually figured out a work around which was to use App_offline.htm and then wait for a minute and hope the files are free.  99% of the time they are, you just have to wait, i guess my app takes a while to power down.

Gone are the days of .net 4.6 where you just copied your dlls over and your app auto recycled.  I guess that was too simple.

**Comment by kanadaj** [comment]
Well yes I guess, but that method isn't very straightforward in a production environment. Without HA your app will end up not responding for a minute, and with HA you get some random bad requests until the load balancer figures out the app is down... Unless you manually cycle the hosts in your load balancer. But that sounds like even more trouble to set up.

**Comment by bladefist** [comment]
@kanadaj yep, we use a load balancer and it linearly updates the servers. No downtime.

**Comment by davidglassborow** [comment]
We have had success with

- Use of a symbolic link for the app directory
- Deploy a new version of the site to a different directory
- Update the symbolic link
- Restart the app pool using the `appcmd` utility that comes with IIS.

**Comment by bladefist** [comment]
@davidglassborow that's a creative solution. Thanks.  Creative solutions shouldn't have to be needed though.

**Comment by davidglassborow** [comment]
In my experience, the old mechanism wasn't reliable when the site was under load anyway, we ended up putting reverse proxies anyway for high traffic sites.

**Comment by davidfowl** [comment]
Right, in-place deployments were never reliable or atomic, if you got lucky or didn't have much traffic it would be fine though. I'm assuming the majority of complains are around those scenarios. We will look into improving things in the .NET 5.0 time frame. Some caveats:
- We will look into making things better for framework dependent applications. If you deploy self contained then you're out of luck. It'll lock both native and managed files in the output folder. The locking native dlls problem has always existed and won't be fixed by whatever mitigation we put in place.
- We plan on looking into loading applications assemblies into memory as bytes instead of loading them from disk and locking the file. This should be fine for most cases but may result in more memory usage some cases.
- In order to recycle the application, you still need to add and remove an appoffline file. Adding the appoffline file is a way to signal that you want the application to restart as a result of the deployment.

**Comment by bladefist** [comment]
@davidfowl The plan you outlined for 5.0 sounds great.  We use framework dependent as I'm sure most do.  

Respectfully disagree about the reliability of in-place deployments.  I've been doing it for years on sites with extremely high load, never had an issue, not once.  The first detected change causes the app to reset, the load balancer sees the instance is not responding, takes it offline, once all the DLLs are replaced, load balancer completes a health check and it comes back online.  Now if there was no load balancer health check, then ya, it's a gamble, but I mean who is replacing DLLs in production under high load w/o a proxy of some kind?

**Comment by davidfowl** [comment]
> Respectfully disagree about the reliability of in-place deployments. I've been doing it for years on sites with extremely high load, never had an issue, not once. The first detected change causes the app to reset, the load balancer sees the instance is not responding, takes it offline, once all the DLLs are replaced, load balancer completes a health check and it comes back online. Now if there was no load balancer health check, then ya, it's a gamble, but I mean who is replacing DLLs in production under high load w/o a proxy of some kind? Lets focus on the rule and not the exception. Without a proxy I cannot think of any way to update the app without causing an outage. Even if you supported hot reloading, some requests would probably hang or fail.

You have a load balancer. Lots of people don't have a load balancer and are just replacing files on disk in place and expecting magic 😄

**Comment by bladefist** [comment]
@davidfowl  Right but high traffic = no lb = reckless 

Also, the issue with app_offline.htm is that from a ci/cd process you have no idea when the process stops.  We had to add a huge sleep command between making that file and pushing the DLLs because our shut down process can take anywhere from 1s to 1 minute.  

Loading the files into memory fixes all that though.

**Comment by davidfowl** [comment]
> @davidfowl Right but high traffic + no lb = reckless

👍 

> Also, the issue with app_offline.htm is that from a ci/cd process you have no idea when the process stops. We had to add a huge sleep command between making that file and pushing the DLLs because our shut down process can take anywhere from 1s to 1 minute. When you're load balancing tons of instances those sleeps add up to very long deploy cycle.

Right, so the problem is that dropping that file doesn't tell you when its ok to start the deployment as files are still locked until the notification is received by ANCM.

**Comment by Tratcher** [comment]
> Also, the issue with app_offline.htm is that from a ci/cd process you have no idea when the process stops.

Right, the VS publisher uses short sleeps and several retries.

**Comment by Socolin** [comment]
> * In order to recycle the application, you still need to add and remove an appoffline file. Adding the appoffline file is a way to signal that you want the application to restart as a result of the deployment.

Would it be possible to have a "apprestart" file to mark the app to be recycled without having to delete the appoffline file afterward ?

**Comment by davidfowl** [comment]
@Socolin why is that?

**Comment by Socolin** [comment]
> @Socolin why is that?

If I understand well (tell me if I'm wrong), what you planning is to remove the lock, but we'll still have to create appoffline to signal IIS to recycle the app after the build is complete?

To do so we'll have to create the file appoffline, then delete it. Can this file be deleted immediately after the file was created ? or should we wait for IIS to detect it? If it's required to wait for a little I would prefer to have a file like "apprestart" that IIS will listen, then delete once it detects it and start recycling.

-----

Also, I have another question in mind. It's about IDE optimization during the build.

Currently, I work on a project that uses dependencies, we have a project for the Web layer, and other project for Logic and Data. When I work in the Web layer (ASP.NET Core) during the build, I create an appoffline file before the build, then delete it after the build and it works.

But when I work in one of the dependencies, when I build them from Rider (I think it's doing the same on VS since it's using msbuild) when I look at what msbuild is doing, once the build of the project is is complete, the dll is copied directly in the .Web/bin directory without rebuilding the .Web project. And since the dll is locked, it silently fail, and I have to manually restart the server or manually trigger the build of the .Web project.

Have you any solution for this?
So far the workaround I have found is to write a small program running in background that look for the /bin of my .Web project. When a file change, it create appoffline file into the directory where IIS listen, then copy the modified files, then delete the appoffline. But it feels a bit hacky but so far it's the only way I've found to be able to build, then press F5 and test my changes.

**Comment by davidfowl** [comment]
> To do so we'll have to create the file appoffline, then delete it. Can this file be deleted immediately after the file was created ? or should we wait for IIS to detect it? If it's required to wait for a little I would prefer to have a file like "apprestart" that IIS will listen, then delete once it detects it and start recycling.

How would IIS know when you’re done copying files? How do you avoid the torn state? You basically need to signalr the starting and completion of a transaction. The creation of the file signals the start and the deletion signals the end.

**Comment by davidfowl** [comment]
> Have you any solution for this?
So far the workaround I have found is to write a small program running in background that look for the /bin of my .Web project. When a file change, it create appoffline file into the directory where IIS listen, then copy the modified files, then delete the appoffline. But it feels a bit hacky but so far it's the only way I've found to be able to build, then press F5 and test my changes.

VS handles this by killing the process before starting a new file for anything that might impact the building of that project.

**Comment by Socolin** [comment]
> > To do so we'll have to create the file appoffline, then delete it. Can this file be deleted immediately after the file was created ? or should we wait for IIS to detect it? If it's required to wait for a little I would prefer to have a file like "apprestart" that IIS will listen, then delete once it detects it and start recycling.
> 
> How would IIS know when you’re done copying files? How do you avoid the torn state? You basically need to signalr the starting and completion of a transaction. The creation of the file signals the start and the deletion signals the end.

If there is no lock, why notify IIS when the build start ? Can't we just signal when all the files are updated and the application need to be recycled ? I think I misunderstood something.

**Comment by davidfowl** [comment]
> If there is no lock, why notify IIS when the build start ? Can't we just signal when all the files are updated and the application need to be recycled ? I think I misunderstood something.

Yes that would work. I thought you were describing the current state of the art. It might even be possible to do this today with web.config. If you touch the file it'll reboot the application.

**Comment by FlukeFan** [comment]
I've had success with in-place deployment (https://flukefan.github.io/ZipDeploy/), by:

1.  Rename assembly files (apparently you can rename even though you can't delete/overwrite);
2.  Copy in the new assemblies;
3. Touch the web-config to restart the app;
4.  Delete the old renamed assemblies.

No load-balancing; single IIS instance; no IIS changes.  YMMV.

**Comment by davidfowl** [comment]
@FlukeFan How do you avoid torn deployments or is that not a concern for you? By torn, I mean that there's a window where the old app loads a new assembly before restarting the application.

**Comment by FlukeFan** [comment]
I don't touch the web.config until all assemblies are renamed/updated.  I guess there is a small chance that something else could cause an app-pool to recycle during that time, but it's not happened yet.

I usually turn off all the other auto-recycle options, and also disable overlapped recycle in the IIS app-pool settings.

**Comment by davidfowl** [comment]
@FlukeFan not a recycle but if you haven't loaded the assembly that got replaced as yet. For example, say you had an assembly A.dll and B.dll. Lets say B is used when you go to the settings page so is loaded then. If you went to the settings page when B.dll was replaced with the new one, you could end up with the old app running with a new B.dll.

**Comment by davidfowl** [comment]
I do wonder if we should make a dotnet global tool to assist in this type of deployment. The tool could basically do all the hard things people have mentioned on this thread. 

Thoughts?

**Comment by FlukeFan** [comment]
So I don't handle that case just now.  https://github.com/FlukeFan/ZipDeploy/blob/master/ZipDeploy/ZipDeploy.cs#L55

I guess I could do things in a different order, touch the web.config first, await all 'other' requests finishing, then finally do the replace just before the last request finished (while IIS queues up new requests for the about-to-recycle app-pool), but what I have works for me just now.

It's probably not currently robust enough for the scenario you're describing (but I haven't been doing anything fancy with dynamic assembly loading other than at startup time).

**Comment by FlukeFan** [comment]
> @FlukeFan not a recycle but if you haven't loaded the assembly that got replaced as yet. For example, say you had an assembly A.dll and B.dll. Lets say B is used when you go to the settings page so is loaded then. If you went to the settings page when B.dll was replaced with the new one, you could end up with the old app running with a new B.dll.

So I don't handle that case just now. https://github.com/FlukeFan/ZipDeploy/blob/master/ZipDeploy/ZipDeploy.cs#L55

I guess I could do things in a different order, touch the web.config first, await all 'other' requests finishing, then finally do the replace just before the last request finished (while IIS queues up new requests for the about-to-recycle app-pool), but what I have works for me just now.

It's probably not currently robust enough for the scenario you're describing (but I haven't been doing anything fancy with dynamic assembly loading other than at startup time).

**Comment by atobrandon** [comment]
> Are there plans to revisit this feature and put it on the road map? It's very inconvenient/unfriendly to require everybody deploying a .Net Core website to manually implement a staging-slot strategy if they want to support zero downtime deployments.
> 
> Having this feature would make the transition to .Net Core websites much smoother for many people, and would allow for faster adoption of .Net Core websites.

I never thought of that. But you may very well have hit the nail on the head. This could be a nefarious business decision to force people into using azure slots. If is was possible in the past why has it taken so long to "fix" now? It's not broken from a business perspective what that in mind. Maybe I'm a little cynical. But I've slowly watched azure features that used to be free whittled away and moved into more expensive tiers.

**Comment by FlukeFan** [comment]
> I do wonder if we should make a dotnet global tool to assist in this type of deployment. The tool could basically do all the hard things people have mentioned on this thread.
> 
> Thoughts?

This is to run on the CI server, say?  How would it push the binaries/package to the destination server?

**Comment by davidfowl** [comment]
> I never thought of that. But you may very well have hit the nail on the head. This could be a nefarious business decision to force people into using azure slots. If is was possible in the past why has it taken so long to "fix" now? It's not broken from a business perspective what that in mind. Maybe I'm a little cynical. But I've slowly watched azure features that used to be free whittled away and moved into more expensive tiers.

In place deployments without coordination are fundamentally unreliable for reasons mentioned above. We’re never going to implement shadow copying again because that feature has historically unreliable and has caused lots of issues in ASP.NET on .NET framework. It also only works for a specific type of deployment, framework dependent which is all .NET Framework supported. .NET core supports self contained and single file which make it much harder to support. Even the suggestion above doesn’t cover those cases.

We still believe doing in place deployments should  be as atomic as possible and lots of the problems on this issue come from bugs and unreliability of app_offline detection and recycle so we spent lots of time fixing that (and pdb locking as well).

Web deploy is the only tool today that embodies the flow with which we think it makes sense to deploy but it’s trapped in that tech so it doesn’t help if you’re ftp deploying to your servers or copying to a file share (though it could).

A dotnet global tool might be an answer here.

> This is to run on the CI server, say? How would it push the binaries/package to the destination server?

I don’t think so. It would probably support FTP and a straight copy. Anything other than that is likely a proprietary protocol.

**Comment by bladefist** [comment]
Hopefully when you say FTP you also mean ftp over ssh.  Such a tool would be very valuable.

**Comment by davidfowl** [comment]
> Hopefully when you say FTP you also mean ftp over ssh. Such a tool would be very valuable.


No I meant ftp. But we would do whatever made sense. Maybe it’s best to be a copy tool that you need to run on the target machine. That gets us out of the business of having to send bits anywhere. 

Also are you SSHing into your windows machines to deploy to IIS?

**Comment by BrightSoul** [comment]
@davidfowl Please add (opt-in) support for deployment slots.
https://github.com/dotnet/aspnetcore/issues/3719#issuecomment-473183712
https://github.com/dotnet/aspnetcore/issues/3793#issuecomment-335666414

> We plan on looking into loading applications assemblies into memory as bytes instead of loading them from disk and locking the file. This should be fine for most cases but **may result in more memory usage some cases**.

Moreover, that's won't let us achieve zero downtime.

**Comment by davidfowl** [comment]
There’s no way we’re going to invest in a feature like that with the module.

**Comment by bladefist** [comment]
> > Hopefully when you say FTP you also mean ftp over ssh. Such a tool would be very valuable.
> 
> No I meant ftp. But we would do whatever made sense. Maybe it’s best to be a copy tool that you need to run on the target machine. That gets us out of the business of having to send bits anywhere.
> 
> Also are you SSHing into your windows machines to deploy to IIS?

Yes.  We use CI/CD and deploy to both linux and windows systems, so it was just a lot easier to install openssh on the windows boxes.

**Comment by zhongqingshen** [comment]
$appPool = 'default'
Stop-WebAppPool -Name $appPool -Passthru
... deploy...

Start-WebAppPool -Name $appPool -Passthru
...deploy.....

or 

$appPool = 'default'
Stop-WebAppPool -Name $appPool 
while((Get-WebAppPoolState -Name $appPool ).Value -ne 'Stop'){
    sleep 1  #second
}
... deploy...

Start-WebAppPool -Name $appPool 
while((Get-WebAppPoolState -Name $appPool ).Value -ne 'Start'){
    sleep 1  #second
}

...deploy.....

**Comment by CJHarmath** [comment]
*Poor mans deployment slots*
About a year ago I’ve posted about creating 2 site root folders and have a symlink point to one of them configured for the IIS site.
The benefit is that your downtime is less comparing to the other simpler solutions.
If overlapping recycle is enabled then no downtime, only warmup penalty for the first request which is the case anyway.

If the deploy/copy operation takes N seconds to complete, you don’t need to stop the app pool for N seconds.
No messing with the questionable app offline.
Just copy to the inactive folder, swap the symlink when the files are there then recycle the app pool.
Atomic deploy, no downtime just slower first request as usual.

https://github.com/dotnet/aspnetcore/issues/3793#issuecomment-459870870

Did anyone try that?
If something like this works then could be turned into a service with a client side deploy tool.

* client -> server whats the inactive folder/share
* S: here
* C: ok deployed / copied
* S: symlink swapped and recycled
* C: thx, http get /healthcheck (optional)
* C: ugh I messed up, rollback pls
* S: symlink swap, recycle (back in business)
* C: rinse repeat until works

**Comment by foxjazz** [comment]
So I am also having issue with deployment locked files. in app hosting. I have tried stopping the app pool, but the files remain locked. I have tried pushing the app_offline.htm but no luck there either.
Also if you want to manage deleting the app_offline.htm file you can easily do that on starting the app in program.cs. This way we know when it starts it becomes online.

The current problem I am facing is that the dll's are hung and even if you rename then copy over the files, the old process still running against the renamed files. So when we attempt to deploy again, we would have to rename the files again. I guess we could use a date of some sort as a file name replacement. 
We are using dotnet core 3.1.4 for this on IIS server windows 2012 R2

**Comment by davidfowl** [comment]
@foxjazz how are the files still locked if the app pool/process is stopped? Can you elaborate?

**Comment by foxjazz** [comment]
> @foxjazz how are the files still locked if the app pool/process is stopped? Can you elaborate?

![image](https://user-images.githubusercontent.com/867496/85180893-d1298900-b241-11ea-8dbb-ec648c402694.png)

![image](https://user-images.githubusercontent.com/867496/85181048-3ed5b500-b242-11ea-94b4-d8b3c67e78cc.png)

When trying to delete the used dll's the picture says used by another process.
The app pool has been offline for a while.

**Comment by bladefist** [comment]
Check task manager to see if the process is still running.  Hitting the stop button doesn't mean it immediately ends.

**Comment by foxjazz** [comment]
> Check task manager to see if the process is still running. Hitting the stop button doesn't mean it immediately ends.

the above picture is from task manager. Not sure which one it is, but if I delete these, it will release the file. I guess I can use handle to figure out the pid to terminate. I didn't think I would have to go to these lengths for deployment.

**Comment by davidfowl** [comment]
That’s unlikely. The file is locked by a process. If w3wp is that process then it was either never killed to begin with or a new one spun up with a different pid locking the same file. That’s why you need to make sure app offline is down first to make sure there’s no new process to lock the file, kill the existing process, deploy new bits, remove app offline

**Comment by foxjazz** [comment]
> That’s unlikely. The file is locked by a process. If w3wp is that process then it was either never killed to begin with or a new one spun up with a different pid locking the same file. That’s why you need to make sure app offline is down first to make sure there’s no new process to lock the file, kill the existing process, deploy new bits, remove app offline

when you say app_offline.htm is down first. Yes this was the case. I have a status path I can check via a get request, and it is definitely offline. Also stopped the app pool manually. But the process w3wp was still hanging the file.

**Comment by davidfowl** [comment]
Then the process wasn’t stopped

**Comment by foxjazz** [comment]
> Then the process wasn’t stopped

So why is it that stopping the app pool, not stop the process?

**Comment by davidfowl** [comment]
The app pool does represent the process and if you stop the app pool it will shit the process down. The file can’t be locked if there’s no process to lock it so one of those other theories seems more likely.

**Comment by foxjazz** [comment]
> The app pool does represent the process and if you stop the app pool it will shit the process down. The file can’t be locked if there’s no process to lock it so one of those other theories seems more likely.

hah you said shit. I don't mean to seem like I am punking you. I am saying that stopping the app pool is NOT working as advertized in all the documentation. The process is not shit down, even after 90 seconds, it is still running. The file is locked even when you can see clearly that app pool has been stopped. The process contains a singleton to get some data. And other controllers. It would be nice to have a script that would shut down the process based on the locked file. I could then deploy without fear. Renaming may work also. Files can be renamed.

**Comment by drewmperkins** [comment]
> About a year ago I’ve posted about creating 2 site root folders and have a symlink point to one of them configured for the IIS site.
> 
> Did anyone try that?
> If something like this works then could be turned into a service with a client side deploy tool.

@CJHarmath - I did give this a try, and have the modified code I'm using below.  Works very well, Thanks!

If I published updates more I might consider a client deploy tool, but really I can just remote shell into the IIS server and execute the powershell script to flip it.

`# Must run as admin
Import-Module WebAdministration
# create 2 site root directories
$a = 'C:\Websites\MySiteA'
$b = 'C:\Websites\MySiteB'
$appRoot  = 'C:\Websites\MySite'
$appName  = 'MyAppName'
$siteName = 'MySiteName'
$poolName = "MyPoolName"
New-Item -Type Directory $a -ErrorAction SilentlyContinue
New-Item -Type Directory $b -ErrorAction SilentlyContinue
# create a symlink to targeting one side
New-Item -Type SymbolicLink -Path $appRoot -Target $a -Name A -ErrorAction SilentlyContinue
New-Item -Type SymbolicLink -Path $appRoot -Target $a -Name B -ErrorAction SilentlyContinue

# point the site root to the symlink
$currentPath = (Get-ItemProperty "IIS:\Sites\$siteName\$appName" -name physicalPath)
if ($currentPath -eq "$appRoot\A") {
    Write-Host "Switched to B"
    New-Item $b\active.txt
    Remove-Item $a\active.txt
    Set-ItemProperty "IIS:\Sites\$siteName\$appName" -name physicalPath -value $appRoot\B
} else {
    Write-Host "Switched to A"
    New-Item $a\active.txt
    Remove-Item $b\active.txt
    Set-ItemProperty "IIS:\Sites\$siteName\$appName" -name physicalPath -value $appRoot\A
}
Restart-WebAppPool -Name $poolName`

**Comment by CJHarmath** [comment]
>  `psexec.exe \\IIS_SERVER cmd /c "powershell -noninteractive -file C:\FlipSymLink.ps1"`

This could also be turned into a [JEA endpoint](https://docs.microsoft.com/en-us/powershell/scripting/learn/remoting/jea/overview), so you can run the `flip` as a non-elevated user (eg. from a CI/CD server deploy step)
`Invoke-Command -ComputerName IIS_SERVER -ConfigurationName IIS-Flip -ScriptBlock { Switch-SymlinkTarget -SiteName MySite}`
Used `Switch` instead of `Flip` as it's an approved verb.

**Comment by peter-dolkens** [comment]
> Used `Switch` instead of `Flip` as it's an approved verb.

`Swap` is the verb used when dealing with slots - would make sense to re-use here it seems?

**Comment by arilishu** [comment]
having the same problem as described 4 years ago.
Any update on this?
I don't want to open Remote desktop to just upload a new dll file.

When I try to overwrite .NET Core application DLL file using FTP on production IIS server file is locked and it cannot be overwritten.

To deploy new version I need to stop application in IIS (opening RDP) to release lock, then overwrite.
It is not possible to deploy without stopping application.


Any solution for this? Thanks

**Comment by LucidObscurity** [comment]
> If you can't take your host offline behind the loadbalancer and would want to do it quicker, here is one way with symlinks and some powershell.
> I was thinking of doing something like this.
> 
> It's quicker as you don't have to stop the app pool during the copy + recycle is better than a hard stop as it lets the current requests complete, so less downtime.
> 
> ```powershell
> # Setup
> Import-Module WebAdministration
> # create 2 site root directories
> $a = 'C:\inetpub\AspNetCoreSampleA'
> $b = 'C:\inetpub\AspNetCoreSampleB'
> $siteRoot = 'C:\inetpub\aspnetcoresample'
> $siteName = 'AspNetCoreSample'
> $poolName = "aspnetcore"
> New-Item -Type Directory $a
> New-Item -Type Directory $b
> # create a symlink to targeting one side
> New-Item -Type SymbolicLink -Path $siteRoot -Target $a
> # point the site root to the symlink
> Set-ItemProperty "IIS:\Sites\$siteName" -name physicalPath -value $siteRoot
> # make sure it get's picked up
> Restart-WebAppPool -Name $poolName
> 
> # this tells you the active side
> Get-Item -Path $siteRoot | Select-Object -ExpandProperty target
> 
> # Flip the symlink
> $current = (Get-Item -Path $siteRoot).Target
> $newTarget = if ($current -eq $a) {$b} else {$a}
> New-Item -Type SymbolicLink -Path $siteRoot -Target $newTarget -Force
> # at this point w3wp.exe still locks the current target folder until it's getting recycled
> # Deploy new version to the symlink which is now pointing to the other side which should have no locks
> robocopy \\myshare\myapp $siteRoot /mir
> # recycle app pool, so it picks up the new files
> Restart-WebAppPool -Name $poolName
> 
> # bonus point: rollback is easy
> $current = (Get-Item -Path $siteRoot).Target
> $newTarget = if ($current -eq $a) {$b} else {$a}
> New-Item -Type SymbolicLink -Path $siteRoot -Target $newTarget -Force
> Restart-WebAppPool -Name $poolName
> ```
> 
> Here is the gist
> https://gist.github.com/csharmath/b2af0f50700ce9fbdd8c5c3e582fd41b

Here's something which I think I've got the kinks worked out. I've only tried with ASP.NET, but I imagine would work the same with the hosting module for net core. The first run does stop the site for a bit, so may want to tweak that a bit.
```powershell
param(
    [Parameter(Mandatory = $true)]
    [string] $iisRootPath,
    [Parameter(Mandatory = $true)]
    [string] $siteName,
    [Parameter(Mandatory = $true)]
    [string] $artifactPath,
    [Parameter(Mandatory = $false)]
    [string] $siteFolderName,
    [Parameter(Mandatory = $false)]
    [string] $appPoolName
)
Import-Module WebAdministration

#populate optional parameters
if (!($PSBoundParameters.ContainsKey('siteFolderName'))) { $siteFolderName = $siteName }
if (!($PSBoundParameters.ContainsKey('appPoolName'))) { $appPoolName = $siteName }

#set A, B and C paths
$a = "$iisRootPath\$siteFolderName" + 'A'
$b = "$iisRootPath\$siteFolderName" + 'B'
$c = "$iisRootPath\$siteFolderName"

#existence test
$cName = Get-Item -Path $c -ErrorAction SilentlyContinue | Select-Object -ExpandProperty name -ErrorAction SilentlyContinue
$bName = Get-Item -Path $b -ErrorAction SilentlyContinue | Select-Object -ExpandProperty name -ErrorAction SilentlyContinue

#get the shudown timeout for the app pool
$shutdownTimeout = Get-WebConfigurationProperty -Filter 'system.web/httpRuntime' -PSPath "IIS:\Sites\$siteName" -Name shutdownTimeout | Select-Object -ExpandProperty Value

#create symlink, if symlink source is an existing directory, rename existing 
if($cName -eq $null) 
{
    Write-Output "Creating SymbolicLink $c -> $a" 
    New-Item -Type SymbolicLink -Path $c -Target $a
} 
else
{
    #directories don't have a target property
    $cTarget = Get-Item -Path $c | Select-Object -ExpandProperty target -ErrorAction SilentlyContinue

    #this is a directory, rename and create symlink
    if($cTarget -eq $null)
    {
        #restart AppPool first so no locked files
        #Write-Output "Restarting AppPool $appPoolName" 
        #Restart-WebAppPool -Name $appPoolName

        #stop AppPool first so no locked files
        Write-Output "Stopping AppPool $appPoolName" 
        Stop-WebAppPool -Name $appPoolName

        #sleep for shutdownTimeout
        Write-Output "Sleeping for $shutdownTimeout" 
        Start-Sleep -Seconds (([System.TimeSpan]::Parse("$shutdownTimeout").TotalSeconds) * 1.1)

        try {
             #if the rename fails, there are files in use. stop the script
            $newName = $siteFolderName + 'A'
            Write-Output "Renaming $c to $newName" 
            Rename-Item -Path $c -NewName $newName -ErrorAction Stop
        }
        catch {
            Write-Error -Message "Failed to rename $c to $newName due to files in use."

            #start AppPool 
            Write-Output "Starting AppPool $appPoolName due to failed folder rename. Consider trying again durring a time with reduced traffic." 
            Start-WebAppPool -Name $appPoolName
            
        }
       
        #create symlink
        Write-Output "Creating SymbolicLink $c -> $a" 
        New-Item -Type SymbolicLink -Path $c -Target $a

        #start AppPool 
        Write-Output "Starting AppPool $appPoolName" 
        Start-WebAppPool -Name $appPoolName

        #copy a to b to pick up directory permissions
        if($bName -eq $null) 
        {
            Write-Output "Copying Directory $a to $b"  
            robocopy $a $b /e /sec /sl
        }
    }
}

#existence test
$aName = Get-Item -Path $a -ErrorAction SilentlyContinue | Select-Object -ExpandProperty name -ErrorAction SilentlyContinue
$bName = Get-Item -Path $b -ErrorAction SilentlyContinue | Select-Object -ExpandProperty name -ErrorAction SilentlyContinue

#create if needed
if($aName -eq $null) 
{
    Write-Output "Creating Directory $a"  
    New-Item -Type Directory $a 
}
if($bName -eq $null) 
{
    Write-Output "Creating Directory $b"  
    New-Item -Type Directory $b 
}

#make sure site pointing to symlink
Write-Output "Pointing Website $siteName to Path $c" 
Set-ItemProperty "IIS:\Sites\$siteName" -name physicalPath -value $c

#restart so we know it's loading from symlink
Write-Output "Restarting AppPool $appPoolName" 
Restart-WebAppPool -Name $appPoolName

#sleep for shutdownTimeout 
Write-Output "Sleeping for $shutdownTimeout" 
Start-Sleep -Seconds ([System.TimeSpan]::Parse("$shutdownTimeout").TotalSeconds)

#get current symlink target and set new target
Get-Item -Path $c | Select-Object -ExpandProperty target
$current = (Get-Item -Path $c).Target
$newTarget = if ($current -eq $a) {$b} else {$a}

$fileOrDirectoryName = Get-Item $artifactPath | Select-Object -ExpandProperty name

#test for .zip
if(($fileOrDirectoryName).ToLower().EndsWith(".zip") -eq $true)
{
    #copy to temp location
    $guid = [System.Guid]::NewGuid()
    $tempPath = "$iisRootPath\temp-$guid"
    $tempDest = "$iisRootPath\temp-dest-$guid"

    Write-Output "Creating temp directory $tempPath" 
    New-Item -Type Directory $tempPath

    $artifactDirectory = [System.IO.Path]::GetDirectoryName($artifactPath)
    $artifactFile = [System.IO.Path]::GetFileName($artifactPath);
    
    #TODO: first robocopy arg needs to be a directory, not a file
    Write-Output "Copying archive from $artifactPath to $tempPath" 
    robocopy $artifactDirectory $tempPath $artifactFile

    #extract to temp destination (seems a failed extract can leave empty folders)
    Write-Output "Extracting archive to $tempDest" 
    Expand-Archive -Path "$tempPath\$fileOrDirectoryName" -DestinationPath $tempDest -Force

    #copy from temp destination to destination
    Write-Output "Copying files from $tempDest to $newTarget" 
    robocopy $tempDest $newTarget /e

    #delete temp directory
    Write-Output "Removing temp directory $tempPath" 
    Remove-Item -Path $tempPath -Recurse -Force

    #delete temp directory
    Write-Output "Removing temp directory $tempDest" 
    Remove-Item -Path $tempDest -Recurse -Force
}
else 
{
    #copy new files
    Write-Output "Copying files from $artifactPath to $newTarget" 
    robocopy $artifactPath $newTarget /e
}

#swap symlink
Write-Output "Pointing SymbolicLink $c -> $newTarget" 
New-Item -Type SymbolicLink -Path $c -Target $newTarget -Force

#restart so it loads from newly swapped symlink
Write-Output "Restarting AppPool $appPoolName" 
Restart-WebAppPool -Name $appPoolName
```

**Comment by arilishu** [comment]
@LucidObscurity what is that? a cmd code? where do i have to add that code?
Thanks

**Comment by joseotavioq** [comment]
What about the idea of increase the number of retries using the parameter `-retryAttempts:X` for the `msdeploy` command, until the process release the file lock?

**Comment by minhhut** [comment]
I think @davidfowl is right, the process still there somehow and locks the files. If you drop app_offline.htm and wait for 90 seconds (default timeout setting to shutdown AppPool), and try again, you will see it works. 
For me on local, I don't want to wait that long so I updated this **Shutdown Time Limit** (seconds) to 5 and update project file to generate **app_offline.htm** before build, remove it after built. It works well. 
For server, I use **msdeploy** to drop app_offline.htm, sync new version, and then remove app_offline.htm. With msdeploy, it does retry if the files are locked but sometime gave up before 90 seconds, so I have to trigger deploy again, then it works at second try

**Comment by arilishu** [comment]
Any update about this? When this new release is gonna be ready?

**Comment by joelmandell** [comment]
I found about this problem this week, after having converting a .NET 4.5.2 application to .NET 5.0. I have another approach to overcome this. I created a Windows Service that is running on the IIS server. On that Service I have a simple HTTP Server running on custom port. The webapplication can do post-request to that server to initiate a upgrade. It stops the app-pool, create backup of current release, removes current release, download upgradepackage and extract it. Then starts app-pool again.

I created a simple Windows Forms application, to create upgrade packages that is pushed to a repo that the above mentioned Windows Service fetches from.

**Comment by shirhatti** [comment]
@jkotalik will have a design proposal for 6.0-preview1

**Comment by jkotalik** [comment]
I have a prototype [here of shadow copying](https://github.com/dotnet/aspnetcore/tree/jkotalik/shadowCopy). This would be an opt-in option to allow running your app from a different temp folder location. This currently would affect first request time though as all files would need to be copied to the directory first before starting the app.

**Comment by jkotalik** [comment]
I've created a PR that implements a version of shadow copying with a design proposal. We'd welcome feedback and people trying this out once available to make sure we deliver here.

**Comment by ScottRFrost** [comment]
I found this issue after having upgraded a Web API project from 4 to 5 (skipping the .NET Core days).  I think we can all agree that not being able to deploy easily is a pretty big problem and a large step backwards in functionality from Framework 4.x.

Is there any chance this PR gets backported to 5.0.x, for release soon, or are we really going to have to wait for .NET 6 a year from now to get this?

**Comment by davidfowl** [comment]
No way. This is a massive feature and not something that we'll back port. It'll be a .NET 6 feature (opt-in) only. When we gain enough confidence from usage then we can decide if it's stable enough to backport.

**Comment by ScottRFrost** [comment]
Understandable, just checking.

So, after reading this thread the current best approach (via RDP or SFTP / etc) is:

1. Upload a app_offline.htm file
2. Upload all app files (everything but the main .dll will generally work on the first upload)
3. Keep spamming "Retry" on uploading the main .dll until IIS realizes the app_offline.htm file is there and shuts down, letting the  .dll file upload
4. Delete the app_offline.htm file

Is that correct?

**Comment by bladefist** [comment]
@ScottRFrost  what I do is similar.  

1. create app_offline.html
2. sleep 20 seconds
3. run taskkill /f /fi "USERNAME eq APPNAMEHERE" /im w3wp.exe
4. copy files

#3 you can play around with to match your setup.  We did this because randomly the app would never shut down after the 20s sleep so we just had to kill it.

**Comment by davidfowl** [comment]
Correct, webdeploy does this by default BTW so if you're using webdeploy, these steps are automated. If you aren't using webdeploy, then you essentially need to mimic those steps you mention above.

Determining when the app_offline has been detected by IIS is the trickiest part. You can attempt a copy and retry with some back off. You can try to make a request to the site and check if the response contains the app_offline.htm content (@jkotalik @shirhatti ).

There are some extremely rare cases when you might need to kill the process because the module should detect the app_offline an proceed to do that.

**Comment by Tratcher** [comment]
Note detecting app_offline isn't the only issue, there can also be code running in the app that prevents it from shutting down in a timely fashion.

**Comment by ScottRFrost** [comment]
> Note detecting app_offline isn't the only issue, there can also be code running in the app that prevents it from shutting down in a timely fashion.

Just ran in to this today during further testing.  Even though the app_offline.htm content was being displayed on our API, the primary .dll was still being locked by IIS.  Even 60 seconds after app_offline content was being displayed, we still couldn't write the .dll.  We can't have the production API offline for > a minute every time we want to deploy a change!

Unfortunately, we are going to have to pull on hold all of our migrations from .NET4 ==> .NET 5 until this IIS deploy bug is fixed, which it sounds like is a year away with .NET 6.

**Comment by davidfowl** [comment]
Do you know why the application wasn't shutting down?

**Comment by ScottRFrost** [comment]
It looked like it was held up by the code still executing.  The only way we were able to write the dll in this case was completely stopping the app pool.  We had experimented with automation to stop the app pool, write the file, then restart the app pool, but sometimes restarting the app pool would error out, leaving the app down.

If you could provide direction on what to check when it gets into this state to determine why it's not shutting down, we could potentially try it on a test environment.

**Comment by joelmandell** [comment]
I use the code below for stopping the pool in our Windows Service that is used for deploying. Using ReadToEnd() seems to do the trick, it waits for the process to stop. The next hook downloads zip-package and unzip in webroot, removes zip-package. And spins up apppool.

Seems to take something between 60-90 seconds. 10 customers are running this in production and I have deployed serveral versions since november. So it is a hassle that it is downtime for 60-90 seconds, because the unzipping only takes like 3 seconds, but I will remove this service when we move to .NET 6.0.

```   
void StopPool(string appPool)
{
  using (Process myProcess = new Process())
  {
      myProcess.StartInfo.FileName = @"C:\Windows\System32\inetsrv\appcmd";
      myProcess.StartInfo.Arguments = $@"stop apppool /apppool.name:""{appPool}""";
      myProcess.StartInfo.UseShellExecute = false;
      myProcess.StartInfo.RedirectStandardOutput = true;
      myProcess.Start();
      string stdout = myProcess.StandardOutput.ReadToEnd();
  }
}
```

**Comment by CJHarmath** [comment]
## SymDeploy
Having an official opt-in feature is nice and kudos to that.

While waiting for that, folks can achieve 100% deploy uptime by just utilizing symlinks which is widely known and understood "thing" in *nix land and can serve the purpose just fine on Windows too.
(Even if you don't care about 100% uptime, give this a try as it can be useful in the future to have this in your tool belt)
I just call this SymDeploy

## Details
The idea is that you point your IIS site to a symlink instead of the target directory of your app so your IIS site config never changes (as you don't want that to cause a disruptive site restart).
While the app is running, the files in the symlink target gets locked, so you can't update that, but you can update the symlink without issues.
So you can deploy a vNext version to a new folder and repoint the symlink to that.
IIS w3wp.exe will still have the open file handles to the dll's in the original symlink target folder. i.e.: your v1 is active at this point.
Then when you recycle the app pool with overlapping recycling enabled, vNext gets loaded by a new w3wp.exe while the old w3wp.exe will finish serving the requests in flight then once done it stops. i.e.: for a short amount of time both v1 and v2 are active a the same time.

The actual symlink swap and pool recycle is just 2 lines of code, so it's not that difficult to understand and maintain.
```
New-Item -Type SymbolicLink -Path $siteRoot -Target $newTarget -Force
Restart-WebAppPool -Name $poolName
```

0 downtime for your customers.

Added bonus: rollback is as easy as reverting back the symlink to v1 then recycle.

## PoC
I had this PoC PowerShell gist posted a while ago just to demonstrate the idea with an A/B folder. Think of it as having a staging and production folder which you can swap:
https://gist.github.com/cjharmath/b2af0f50700ce9fbdd8c5c3e582fd41b
As I still see folks keep fighting with app offline which is a sub-optimal solution when you **don't want to have downtime** (just like in the old days) I wanted to post an updated comment.

Here is a diagram for the more visuals in hope that it might better click:
![SymlinkDeploy](https://user-images.githubusercontent.com/1450460/103033524-99710600-4530-11eb-8a23-f8f44c4911b3.png)

## Why symlink and not just update the site root path?
Updating the site root path will bring down your site, so you won't have 100% uptime.

All the other solutions to me seem sub-optimal. Once you get into automating app offline, why not just do this instead and not worry about your downtime? 

## CI/CD pipeline
* add the symlink deploy script to a [PowerShell JEA](https://docs.microsoft.com/en-us/powershell/scripting/learn/remoting/jea/overview?view=powershell-7.1) endpoint on your IIS servers. Polish up the PoC then expose it as a cmdlet on a JEA endpoint and done.
* have your deploy step copy the new version then call the symlink swap + recycle JEA command
* enjoy your 0 downtime continuously deployed aspnet core app with legacy IIS setup

**Comment by joelmandell** [comment]
@CJHarmath It still has file locks. I tried to implement the outlined steps in C# and using appcmd through Process class.
But it does not work. And using the script pointed in the gist produces IOException, ResourceExists.

**Comment by CJHarmath** [comment]
> 
> 
> @CJHarmath It still has file locks. I tried to implement the outlined steps in C# and using appcmd through Process class.
> But it does not work. And using the script pointed in the gist produces IOException, ResourceExists.

I've looked at your comments / screenshot on the gist. you need to add checks for folder exists before trying to create them. 
Also, when updating a symlink with `New-Item -Type SymbolicLink` you need to use the -Force flag.

As per locking, yes it will still lock the dll's in the symlink target folder but that's fine. you can deploy to the other side and swap the symlink to point over there, then recycle.

**Comment by elidexter** [comment]
Just putting into folder like classic asp net /mvc, copy y and update site without stop IIS is the best option.. is very very ugly and annoying to stop  IIS every moment when you need to update... is not functional at all

**Comment by arilishu** [comment]
This means, .net core is not ready for critical apps?

**Comment by elidexter** [comment]
It works good .. but dll of app is locked by IIS when you try to update your app... you need to stop IIS  to complete the update of site.. you know this problem is  not present on Net framework you just update the site without stoping the site.. if you have 10 web servers it going to lag your time.. as you can see is not functional for this case for me is not functional

**Comment by sln162** [comment]
We always use to modify the name of the existing DLL, for example: modify a.dll to 1.dll, then copy the new a.dll to the current location, and then use the recycling function of IIS to complete the "replacement".

**Comment by glenn2223** [comment]
I personally use a PowerShell command. It publishes to a new subdirectory and updates IIS's physical path to the new location. I created a quick repo with the code => [.NET No Lock Deployment](https://github.com/glenn2223/NET-Core-No-Lock-Deployment)

I use VS Code and deploy to IIS using a task (examples in the repo)

---

Could this work with v6 where you would copy the existing to a temp directory, then update the actual directory and switch the path back again?

**Comment by FlukeFan** [comment]
> We always use to modify the name of the existing DLL, for example: modify a.dll to 1.dll, then copy the new a.dll to the current location, and then use the recycling function of IIS to complete the "replacement".
> The online service does not stop, but the connection of single digit will be lost when recycling, and the CPU usage will double, which is within our expectation.

That's what I'm doing above in comment:  https://github.com/dotnet/aspnetcore/issues/3793#issuecomment-583864579

**Comment by UweKeim** [comment]
We do it in our upload/deployment scripts by creating an "App_Offline.htm" file in the target folder, then deploy the files, then rename the "App_Offline.htm" to "_App_Offline.htm" or delete it again.

**Comment by naveedahmed1** [comment]
@UweKeim we do the same, but:

1. Placing `app_offline.html` still keep the `dll` files locked for few seconds.
2. And  to delete `app_offline.htm` we have to rename it, otherwise it doesn't allow us to delete the file.

**Comment by joelmandell** [comment]
> > We always use to modify the name of the existing DLL, for example: modify a.dll to 1.dll, then copy the new a.dll to the current location, and then use the recycling function of IIS to complete the "replacement".
> > The online service does not stop, but the connection of single digit will be lost when recycling, and the CPU usage will double, which is within our expectation.
> 
> That's what I'm doing above in comment:  https://github.com/dotnet/aspnetcore/issues/3793#issuecomment-583864579

This seemed to be the most easiest to implement. It is almost no downtime.

**Comment by Jtech42** [comment]
The element 'PropertyGroup' in namespace 'http://schemas.microsoft.com/developer/msbuild/2003' has invalid child element 'EnableMSDeployAppOffline'

**Comment by ScottRFrost** [comment]
Mentioned in another thread, but there's a 3 month old Pull Request https://github.com/dotnet/aspnetcore/pull/28357 out that that will hopefully fix this deployment experience regression once it's merged.  It received a few reviews a couple of weeks ago.

**Comment by davidfowl** [comment]
Yes @jkotalik is working on this for .NET 6

**Comment by arilishu** [comment]
My solution was deploying directly from visual studio using web deploy to
the server.

**Comment by jkotalik** [comment]
Merged the first part of this PR for the preview3 release. Will add documentation on trying it out soon.

We'd appreciate trying this feature out such that we have ample feedback before trying to ship this feature in .NET 6.

**Comment by BrightSoul** [comment]
Thanks @jkotalik, any improvement on this issue is deeply appreciated. And also thanks for the amount of effort you put into it.
Here's my feedback based on your merge https://github.com/dotnet/aspnetcore/pull/28357

> During this sequence, ANCM will copy the contents of the app directory to another directory which is user specified. The path to this directory can be absolute or relative. [...] We need to copy dlls to a different directory than what is currently running in the app. So in the case where the directory name is '0', we will create a directory name '1' and write the contents there. Then on app start, it will pick the directory name '1' as its the highest value.

You basically did the work needed to implement deployment slots, without actually exposing it as a deployment slots feature. That must be because @davidfowl [said](https://github.com/dotnet/aspnetcore/issues/3793#issuecomment-585966840):

> There’s no way we’re going to invest in a feature like that with the module.

So instead you came up with a more convoluted solution which relies on a non-configurable debouncing timer.
I don't want to sound ungrateful so I'll say "thanks" once more but it's hard to imagine what the reasoning behind this must have been. I'm genuinely interested in knowing though.

**Comment by jkotalik** [comment]
We are still iterating through this feature. We may not even stick with a debouncing method for determining steady state in the future; it's what seemed best at the time.

**Comment by jkotalik** [comment]
@BrightSoul what behavior do you ideally seek here? The other "mode" we are going to look into is a manual restart mode (which I believe is more similar to your deployment slot mode). Effectively, the file watcher won't be looking for dll changes, but all dlls will still be unlocked in your app directory. Then, when you want to redeploy your app, you can add app_offline.htm, which will cause the app to shutdown, and the next time the app starts, the content will be recopied.

**Comment by freemstr** [comment]
@jkotalik 
With app_offline.htm approach would this mean that once the hosting bundle on the IIS server is updated to latest release there would not be changes required to the project and legacy projects would be deployed with azure pipeline in the same way they are now with app_offline.htm is placed in the directory?
If so, it would be very appreciated.

**Comment by arilishu** [comment]
In my case (and I solved this problem) is to use Web Deploy instead of
publishing through FTP without problems!

**Comment by BrightSoul** [comment]
> The other "mode" we are going to look into is a manual restart mode (which I believe is more similar to your deployment slot mode).

In my case a manual restart would be preferrable. I can't risk my application going live with half the content in case the transfer aborted midway due a connectivity issue. How can a debounce timer determine whether all content was correctly delivered or not?

I think a manual way of switching is needed, either by updating the web.config file or by uploading some other ad-hoc file. I'd be the one to tell it "ok, now you can switch" but only after all files have been transferred..

**Comment by sln162** [comment]
@BrightSoul   
I don't like manual restart. I like direct replacement. If multiple files need to be replaced, I will choose to stop the service, switch traffic to other servers, etc.

Of course, if you can web.config There is an option in, which is also acceptable

**Comment by davidfowl** [comment]
> I think a manual way of switching is needed, either by updating the web.config file or by uploading some other ad-hoc file. I'd be the one to tell it "ok, now you can switch" but only after all files have been transferred..

Why not use appoffline then?

**Comment by BrightSoul** [comment]
> Why not use appoffline then?

Because I'd like to achieve 0 downtime.

**Comment by neoffer** [comment]
+1 for 0 downtime deployment.

Also please make .net apps always "always on" and no need to warm-up for better consistent performance.

Please provide separate small package web api and web api controller. Nowadays we don't use MVC anymore, we are using web api and grpc, make these faster.

Make .Net Core faster and smaller than Rust and Golang.

**Comment by davidfowl** [comment]
Honestly this isn't a job for the ANCM module, this is a job of a deployment system. I don't understand how zero downtime deployments are a concern when using web deploy to blast a new copy of a website directly on top of the old one while it's running. 

We're doing this feature because ASP.NET on .NET Framework had a similar unreliable story, this is no different to that and will ease migration. That's what this is.

> Please provide separate small package web api and web api controller. Nowadays we don't use MVC anymore, we are using web api and grpc, make these faster.

Please file an issue with actual benchmarks and scenarios where the performance isn't adequate.

**Comment by jkotalik** [comment]
With preview3 shipping today, we have shipped our first iteration of ShadowCopy support in ANCM. See https://devblogs.microsoft.com/aspnet/asp-net-core-updates-in-net-6-preview-3/#shadow-copying-in-iis for more details.

We'd really appreciate feedback to make sure we are delivering a feature that people want and any bugs that people may hit.

Again, we aren't trying to achieve zero down time deployment with this feature, the main goal is to avoid file locking when deploying an application and avoid the need for app_offline.

**Comment by ahdung** [comment]
Any news? I really expect this.

**Comment by BrightSoul** [comment]
lol @msftbot you're doing more harm than good.

**Comment by davidfowl** [comment]
Bad @msftbot  !

**Comment by Tratcher** [comment]
> Any news? I really expect this.

> We'd really appreciate feedback to make sure we are delivering a feature that people want and any bugs that people may hit.

**Comment by ScottRFrost** [comment]
The shadow copying in IIS has been a beta feature for 4 months now (since preview 3) https://devblogs.microsoft.com/aspnet/asp-net-core-updates-in-net-6-preview-3/#shadow-copying-in-iis

How has feedback / testing been?   Any word / commitments on if it'll be included as a (non-experimental) feature in .NET 6 final this fall?

**Comment by Tratcher** [comment]
> How has feedback / testing been? Any word / commitments on if it'll be included as a (non-experimental) feature in .NET 6 final this fall?

There has been almost no feedback on the feature so far. Since we're now at the RC1 deadline we're planning to leave it as experimental for this release in hopes of getting more feedback.

**Comment by jgasiorowski** [comment]
Hello @jkotalik, I've just used shadow copy feature. I am not sure if I am doing it right. But from first time I was using dotnet core I was trying to set dev environment to work similar to what I've experienced with old asp.net.
Since aspnetcoremodulev2 for IIS I am using in proc hosting which points to root directory of my asp.net core application. And obviously when website was running any try to build website was showing errors about locked files in /bin folder. 
Only way to work with that was to use visual studio which is able to stop->deploy->start apppool. 

Today I read about that shadowcopy feature and got really excited to try it. I've installed new hosting bundle and tried to run it with application using net core 3.1. I can see in logs that shadow copy kicks in (it created new folder with copy of the whole web project contents). But the problem is that I am still getting the same locking error during build if applicaiton is still running

Am I using it wrong? I couldn't find any end to end example how it should work and what is the expected behaviour. Any advice appreciated :)

**Comment by adityamandaleeka** [comment]
@HaoK @shirhatti ^

**Comment by HaoK** [comment]
@jgasiorowski what versions of asp.net core are you using, and what does your web.config look like for the app after publish?

**Comment by Tratcher** [comment]
Please test 6.0 as well. I don't know if this feature is intended to work with prior versions.

**Comment by jgasiorowski** [comment]
@HaoK Full log looks like that
```
[2021-10-09T19:50:22.921Z, PID: 81680] [aspnetcorev2.dll] Initializing logs for 'C:\Program Files\IIS\Asp.Net Core Module\V2\aspnetcorev2.dll'. Process Id: 81680. File Version: 16.0.21245.0. Description: IIS ASP.NET Core Module V2. Commit: 0621c1748eac08dd516edc950559ab7e8308f214.
[2021-10-09T19:50:22.923Z, PID: 81680] [aspnetcorev2.dll] Copying to shadow copy directory F:\temp\iis\shadowcopy\ShadowCopyDirectory\0.
[2021-10-09T19:50:22.934Z, PID: 81680] [aspnetcorev2.dll] Finished copying 13 files to shadow copy directory F:\temp\iis\shadowcopy\ShadowCopyDirectory\.
[2021-10-09T19:50:22.935Z, PID: 81680] [aspnetcorev2.dll] Resolving hostfxr parameters for application: 'dotnet' arguments: '.\test1.dll' path: 'F:\temp\iis\shadowcopy\ShadowCopyDirectory\0'
[2021-10-09T19:50:22.936Z, PID: 81680] [aspnetcorev2.dll] Known dotnet.exe location: ''
[2021-10-09T19:50:22.937Z, PID: 81680] [aspnetcorev2.dll] Process path 'dotnet.exe' is dotnet, treating application as portable
[2021-10-09T19:50:22.938Z, PID: 81680] [aspnetcorev2.dll] Resolving absolute path to dotnet.exe from 'dotnet.exe'
[2021-10-09T19:50:22.939Z, PID: 81680] [aspnetcorev2.dll] Invoking where.exe to find dotnet.exe
[2021-10-09T19:50:23.339Z, PID: 81680] [aspnetcorev2.dll] where.exe invocation returned: 'C:\Program Files\dotnet\dotnet.exe
C:\Program Files (x86)\dotnet\dotnet.exe
'
[2021-10-09T19:50:23.341Z, PID: 81680] [aspnetcorev2.dll] Current process bitness type detected as isX64=1
[2021-10-09T19:50:23.342Z, PID: 81680] [aspnetcorev2.dll] Processing entry 'C:\Program Files\dotnet\dotnet.exe'
[2021-10-09T19:50:23.343Z, PID: 81680] [aspnetcorev2.dll] Binary type 6
[2021-10-09T19:50:23.344Z, PID: 81680] [aspnetcorev2.dll] Found dotnet.exe via where.exe invocation at 'C:\Program Files\dotnet\dotnet.exe'
[2021-10-09T19:50:23.345Z, PID: 81680] [aspnetcorev2.dll] Resolving absolute path to hostfxr.dll from 'C:\Program Files\dotnet\dotnet.exe'
[2021-10-09T19:50:23.346Z, PID: 81680] [aspnetcorev2.dll] hostfxr.dll located at 'C:\Program Files\dotnet\host\fxr\6.0.0-rc.1.21451.13\hostfxr.dll'
[2021-10-09T19:50:23.347Z, PID: 81680] [aspnetcorev2.dll] Converted argument '.\test1.dll' to 'F:\temp\iis\shadowcopy\ShadowCopyDirectory\0\.\test1.dll'
[2021-10-09T19:50:23.348Z, PID: 81680] [aspnetcorev2.dll] Parsed hostfxr options: dotnet location: 'C:\Program Files\dotnet\dotnet.exe' hostfxr path: 'C:\Program Files\dotnet\host\fxr\6.0.0-rc.1.21451.13\hostfxr.dll' arguments:
[2021-10-09T19:50:23.349Z, PID: 81680] [aspnetcorev2.dll] Argument[0] = 'C:\Program Files\dotnet\dotnet.exe'
[2021-10-09T19:50:23.350Z, PID: 81680] [aspnetcorev2.dll] Argument[1] = 'F:\temp\iis\shadowcopy\ShadowCopyDirectory\0\.\test1.dll'
[2021-10-09T19:50:23.351Z, PID: 81680] [aspnetcorev2.dll] Loading hostfxr from location C:\Program Files\dotnet\host\fxr\6.0.0-rc.1.21451.13\hostfxr.dll
[2021-10-09T19:50:23.410Z, PID: 81680] [aspnetcorev2.dll] Canceling standard stream pipe reader
[2021-10-09T19:50:23.411Z, PID: 81680] [aspnetcorev2.dll] Loading request handler:  'C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App\5.0.10\aspnetcorev2_inprocess.dll'
[2021-10-09T19:50:23.413Z, PID: 81680] [aspnetcorev2.dll] Creating handler application
[aspnetcorev2_inprocess.dll] Initializing logs for 'C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App\5.0.10\aspnetcorev2_inprocess.dll'. Process Id: 81680.. File Version: 15.0.21236.10. Description: IIS ASP.NET Core Module V2 Request Handler. Commit: b92c9f50f0333ea661785ba849f303cde6a93844.
[aspnetcorev2_inprocess.dll] Waiting for initialization
[aspnetcorev2_inprocess.dll] Starting in-process worker thread
[aspnetcorev2_inprocess.dll] Resolving hostfxr parameters for application: 'dotnet' arguments: '.\test1.dll' path: 'F:\temp\iis\shadowcopy\publish\'
[aspnetcorev2_inprocess.dll] Known dotnet.exe location: 'C:\Program Files\dotnet\dotnet.exe'
[aspnetcorev2_inprocess.dll] Process path 'dotnet.exe' is dotnet, treating application as portable
[aspnetcorev2_inprocess.dll] Resolving absolute path to hostfxr.dll from 'C:\Program Files\dotnet\dotnet.exe'
[aspnetcorev2_inprocess.dll] hostfxr.dll located at 'C:\Program Files\dotnet\host\fxr\6.0.0-rc.1.21451.13\hostfxr.dll'
[aspnetcorev2_inprocess.dll] Converted argument '.\test1.dll' to 'F:\temp\iis\shadowcopy\publish\.\test1.dll'
[aspnetcorev2_inprocess.dll] Parsed hostfxr options: dotnet location: 'C:\Program Files\dotnet\dotnet.exe' hostfxr path: 'C:\Program Files\dotnet\host\fxr\6.0.0-rc.1.21451.13\hostfxr.dll' arguments:
[aspnetcorev2_inprocess.dll] Argument[0] = 'C:\Program Files\dotnet\dotnet.exe'
[aspnetcorev2_inprocess.dll] Argument[1] = 'F:\temp\iis\shadowcopy\publish\.\test1.dll'
[aspnetcorev2_inprocess.dll] Setting environment variable ASPNETCORE_IIS_HTTPAUTH=anonymous;
[aspnetcorev2_inprocess.dll] Setting environment variable ASPNETCORE_IIS_PHYSICAL_PATH=F:\temp\iis\shadowcopy\publish\
[aspnetcorev2_inprocess.dll] Loading hostfxr from location C:\Program Files\dotnet\host\fxr\6.0.0-rc.1.21451.13\hostfxr.dll
[aspnetcorev2_inprocess.dll] Initial Dll directory: '', current directory: 'c:\windows\system32\inetsrv'
[aspnetcorev2_inprocess.dll] Setting dll directory to c:\windows\system32\inetsrv
[aspnetcorev2_inprocess.dll] Setting current directory to F:\temp\iis\shadowcopy\publish\
[aspnetcorev2_inprocess.dll] In-process callbacks set
[aspnetcorev2_inprocess.dll] Event Log: 'Application 'F:\temp\iis\shadowcopy\publish\' started successfully.' 
End Event Log Message.
[aspnetcorev2_inprocess.dll] Starting app_offline monitoring in application 'F:\temp\iis\shadowcopy\publish\'
[aspnetcorev2_inprocess.dll] Adding request. Total Request Count 1
[aspnetcorev2_inprocess.dll] Starting file watcher thread
[aspnetcorev2_inprocess.dll] Removing Request 0
[2021-10-09T19:53:24.258Z, PID: 81680] [aspnetcorev2.dll] ASPNET_CORE_GLOBAL_MODULE::OnGlobalStopListening
[2021-10-09T19:53:24.261Z, PID: 81680] [aspnetcorev2.dll] Stopping application '/LM/W3SVC/21/ROOT'
[aspnetcorev2_inprocess.dll] Stopping CLR
[aspnetcorev2_inprocess.dll] Drained all requests, notifying managed.
[aspnetcorev2_inprocess.dll] Starting shutdown sequence 0
[aspnetcorev2_inprocess.dll] Waiting for 0 requests to drain
[aspnetcorev2_inprocess.dll] Managed application exited with code 0
[aspnetcorev2_inprocess.dll] Clr thread wait ended: clrThreadExited: 1
[aspnetcorev2_inprocess.dll] Event Log: 'Application 'MACHINE/WEBROOT/APPHOST/TESTSHADOW' has shutdown.' 
End Event Log Message.
[aspnetcorev2_inprocess.dll] Canceling standard stream pipe reader
[aspnetcorev2_inprocess.dll] Stopping in-process worker thread
[aspnetcorev2_inprocess.dll] Stopping file watcher thread
[2021-10-09T19:53:25.402Z, PID: 81680] [aspnetcorev2.dll] ASPNET_CORE_GLOBAL_MODULE::Terminate
```
`Stopping file watcher` log entry was only generated when I was stopping apppool manually


> When you say you try to delete the bin directory

Yes I tried delete whole directory. I checked just now if I can delete only the `test1.dll` file from published location without luck 
![image](https://user-images.githubusercontent.com/5520607/136672482-1f5b6bc2-9ed5-415d-833d-1d60dd44692c.png)
Sorry - it is polish Windows instance. Modal says that file is in use by `IIS Worker Process`

@Tratcher I will try it soon since I don't have sdk for net 6 installed. I've installed only hosting bundle to have iis module available. I will return with the answer. But Based on module code and behaviour it really looks like the wrong path is being locked

**Comment by joelmandell** [comment]
Hi. I just want to say thank you for adding this. This made my life much easier. This feature has been working great for me since its appearance in previews @jkotalik

**Comment by ahdung** [comment]
@joelmandell Introduced in VS2022?

**Comment by ScottRFrost** [comment]
@joelmandell Can you details the steps you did to enable it, and how you deploy now that you got it working?
